lib/linked_list.rb

```diff
@@ -12,150 +12,265 @@ def initialize(value, next_node = nil)
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    last_index = length() - 1
+    return get_at_index(last_index - n)
+  end
+
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(n), a new array is created whose size is dependent on the number of nodes in the list
+  def has_cycle
+    return if !@head
+    current = @head
+    visited = []
+
+    while current
+      return true if visited.include?(current.next)
+      visited.push(current)
+      current = current.next
     end
 
-    # Helper method for tests
-    # Creates a cycle in the linked list for testing purposes
-    # Assumes the linked list has at least one node
-    def create_cycle
-      return if @head == nil # don't do anything if the linked list is empty
+    return false
+  end
+
 
-      # navigate to last node
-      current = @head
-      while current.next != nil
-          current = current.next
+  # Additional Exercises 
+  # returns the value in the first node
+  # returns nil if the list is empty
+  # Time Complexity: O(1), only looking at @head
+  # Space Complexity: O(1)
+  def get_first
+    return @head ? @head.data : nil
+  end
+
+  # method that inserts a given value as a new last node in the linked list
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def add_last(value)
+    return add_first(value) if !@head
+  
+    current = @head
+    current = current.next until !current.next
+    current.next = Node.new(value)
+  end
+
+  # method that returns the value of the last node in the linked list
+  # returns nil if the linked list is empty
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def get_last
+    return if !@head
+    current = @head
+    current = current.next until !current.next
+    return current.data
+  end
+
+  # method to insert a new node with specific data value, assuming the linked
+  # list is sorted in ascending order
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def insert_ascending(value)
+    return if !@head
+    current = @head
+    previous = nil
+
+    while current
+      if current.data > value
+        new_node = Node.new(value, current)
+      elsif !current.next # if we reach the end but haven't found a larger value
+        new_node = Node.new(value)
       end
 
-      current.next = @head # make the last node link to first node
+      previous ? previous.next = new_node : @head = new_node # if previous is nil, the new_node must become the head
+      previous = current
+      current = current.next
+    end

```

Comment:
I don't think this quite works, if `current.data < value` and there are several items in the list, then you may end up removing several nodes by `previous.next = new_node` when `new_node` is `nil`.